File: Stack/0678. Valid Parenthesis String/0678.py
class Solution:
    def checkValidString(self, s: str) -> bool:
        low=high=0
        
        for c in s:
            if c=='(':
                low,high=low+1,high+1
            elif c==')':
                low,high=low-1,high-1
            elif c=='*':
                low,high=low-1,high+1
            
            if high<0:
                return False
            low=max(0,low)
        
        return low==0
            

File: Stack/0678. Valid Parenthesis String/0678.cpp
#pragma GCC optimize('0fast', 'unroll-loops')

class Solution {
public:
    bool checkValidString(string s) {
        ios_base::sync_with_stdio(0);
        cout.tie(0);
        cin.tie(0);

        int low=0;
        int high=0;

        for(char c : s){
            if(c=='('){
                low++;
                high++;
            } else if(c==')'){
                low--;
                high--;
            } else if(c=='*'){
                low--;
                high++;
            }
            if(high<0) return false;
            low=max(0,low);
        }
        return low==0;
    }
};